### PR TITLE
fix(overlay): fixes keyboard navigation and focus management issues

### PR DESCRIPTION
**Pull request by cdransf**
## Description

Fixes accessibility issue where keyboard users could not navigate into interactive content within hover-triggered overlays. When a user tabbed to a hover trigger and attempted to tab into the overlay content, it would close immediately, making interactive elements (buttons, links, action groups) inaccessible.

This PR extends `HoverController` to track focus within both the trigger and overlay content ("hover interaction zone"), applying the same 300ms delay mechanism for keyboard interactions as pointer interactions. Now keyboard users can successfully tab into and interact with hover overlay content, and press Escape to close and return focus. All changes follow W3C WAI guidelines for fly-out menus.

## Related issue(s)

-   fixes `SWC-1268`

## Screenshots

https://github.com/user-attachments/assets/d61e913a-03d2-46da-a733-437fdc9c1455

## Author's checklist

-   [X] I have read the **[CONTRIBUTING](<(https://github.com/adobe/spectrum-web-components/blob/main/CONTRIBUTING.md)>)** and **[PULL_REQUESTS](<(https://github.com/adobe/spectrum-web-components/blob/main/PULL_REQUESTS.md)>)** documents.
-   [X] I have reviewed at the Accessibility Practices for this feature, see: [Aria Practices](https://www.w3.org/TR/wai-aria-practices/)
-   [X] I have added automated tests to cover my changes.
-   [ ] I have included a well-written changeset if my change needs to be published.
-   [ ] I have included updated documentation if my change required it.

---

## Reviewer's checklist

-   [ ] Includes a Github Issue with appropriate flag or Jira ticket number without a link
-   [ ] Includes thoughtfully written changeset if changes suggested include `patch`, `minor`, or `major` features
-   [ ] Automated tests cover all use cases and follow best practices for writing
-   [ ] Validated on all supported browsers
-   [ ] All VRTs are approved before the author can update Golden Hash

### Manual review test cases

- [ ] Keyboard navigation into hover content with interactive buttons

1. [Go to Storybook → Overlay → Hover With Interactive Content](https://swcpreviews.z13.web.core.windows.net/pr-5804/docs/storybook/?path=/story/overlay--hover-with-interactive-content)
2. Press Tab to focus "Hover for interactive buttons"
3. Wait for popover to appear
4. Press Tab again to enter the popover
5. Expected: Popover stays open, focus moves to "Action 1" button
6. Press Tab through all 3 buttons
7. Expected: Popover remains open while navigating
8. Press Tab to exit the popover
9. Expected: Popover closes after ~300ms delay

- [ ] Keyboard navigation with arrow keys in action group

1. [Go to Storybook → Overlay → Hover With Interactive Content](https://swcpreviews.z13.web.core.windows.net/pr-5804/docs/storybook/?path=/story/overlay--hover-with-interactive-content)
2. Press Tab to focus "Hover for action group"
3. Wait for popover to appear
4. Press Tab to enter the action group
5. Use ↑/↓ arrow keys to navigate between options
6. Expected: Popover stays open during arrow key navigation
7. Press Tab to exit
8. Expected: Popover closes after delay

- [ ] Escape key closes overlay and returns focus

- [Go to Storybook → Overlay → Hover With Interactive Content](https://swcpreviews.z13.web.core.windows.net/pr-5804/docs/storybook/?path=/story/overlay--hover-with-interactive-content)
- Press Tab to focus any trigger button
- Press Tab to enter the popover content
- Press Escape
- Expected: Popover closes immediately and focus returns to trigger button

- [ ] Mouse hover still works correctly

1. [Go to Storybook → Overlay → Hover With Interactive Content](https://swcpreviews.z13.web.core.windows.net/pr-5804/docs/storybook/?path=/story/overlay--hover-with-interactive-content)
2. Hover mouse over "Hover for interactive links"
3. Move mouse into the popover content
4. Expected: Popover stays open
5. Move mouse away from both trigger and popover
6. Expected: Popover closes after ~300ms delay

- [ ] Mixed mouse and keyboard interactions

1. [Go to Storybook → Overlay → Hover With Interactive Content](https://swcpreviews.z13.web.core.windows.net/pr-5804/docs/storybook/?path=/story/overlay--hover-with-interactive-content)
2. Hover mouse over trigger to open popover
3. Press Tab to move focus into popover content
4. Expected: Popover stays open
5. Press Tab to exit popover
6. Expected: Popover closes after delay even though mouse may still be nearby

- [ ] Existing hover tooltip behavior unchanged

1. Go to Storybook → Overlay → any existing story with tooltips (e.g., [Triggered By Optimization](https://swcpreviews.z13.web.core.windows.net/pr-5804/docs/storybook/?path=/story/overlay--triggered-by-optimization))
2. Test both mouse hover and keyboard focus
3. Expected: Tooltips appear and disappear as before (no regression)

### CODE REVIEW

File: packages/overlay/src/HoverController.ts [hunk 1]
```diff
@@ -142,12 +174,34 @@ export class HoverController extends InteractionController {
...
         if (this.focusedin && triggerElement.matches(':focus-visible')) return;
+        // Don't close if focus is within overlay content
+        if (this.overlayFocusedin) return;
 
         this.hoverTimeout = setTimeout(() => {
             this.open = false;
         }, HOVER_DELAY);
     }
 
+    protected doFocusleave(): void {
```
Comment (cdransf) [inline packages/overlay/src/HoverController.ts]: This relies on timeouts and delays more than I would like, but it _does_ adhere to the prescribed `300ms` delay.

File: packages/overlay/src/HoverController.ts [hunk 2]
```diff
@@ -33,9 +33,21 @@ export class HoverController extends InteractionController {
 
     pointerentered = false;
 
+    overlayFocusedin = false;
+
     handleKeyup(event: KeyboardEvent): void {
-        if (event.code === 'Tab' || event.code === 'Escape') {
+        if (event.code === 'Tab') {
```
Comment (cdransf) [inline packages/overlay/src/HoverController.ts:37]: Previously, hitting escape had the same effect as tab — setting open to `true`, which is not consistent with what a user would expect. This change closes the overlay and returns focus to the trigger.

File: packages/overlay/src/HoverController.ts [hunk 3]
```diff
@@ -88,6 +102,24 @@ export class HoverController extends InteractionController {
...
+        this.overlayFocusedin = true;
+        // Clear any pending close timeout when focus enters overlay
+        if (this.hoverTimeout) {
+            clearTimeout(this.hoverTimeout);
+            this.hoverTimeout = undefined;
+        }
+    }
+
+    handleOverlayFocusout(): void {
+        this.overlayFocusedin = false;
```
Comment (cdransf) [inline packages/overlay/src/HoverController.ts]: We have `overlayFocusedIn` (which this PR adds), `pointerentered` and `focusedin` which feel similar semantically, but are used differently.

File: packages/overlay/test/overlay-trigger-hover.test.ts [hunk 4]
```diff
@@ -372,4 +372,292 @@ describe('Overlay Trigger - Hover', () => {
 
         expect(button3 === document.activeElement).to.be.true;
     });
+    describe('keyboard navigation into hover content', () => {
```
Comment (nikkimk) [inline packages/overlay/test/overlay-trigger-hover.test.ts:595]: Thanks for also adding a regression test!

File: packages/overlay/stories/overlay.stories.ts [hunk 5]
```diff
@@ -1654,6 +1655,88 @@ export const triggeredByOptimization = (): TemplateResult => {
     `;
 };
 
+export const hoverWithInteractiveContent = (): TemplateResult => {
```
Comment (nikkimk) [inline packages/overlay/stories/overlay.stories.ts:1658]: Thanks for adding stories to this.

File: packages/overlay/src/HoverController.ts [hunk 6]
```diff
@@ -198,5 +252,20 @@ export class HoverController extends InteractionController {
...
+        );
+        this.overlay.addEventListener(
+            'focusout',
+            () => this.handleOverlayFocusout(),
+            { signal }
+        );
+        this.overlay.addEventListener(
+            'keyup',
+            (event) => this.handleKeyup(event),
+            { signal }
```
Comment (Rajdeepc) [inline packages/overlay/src/HoverController.ts:259]: Great for automatic cleanup. Thanks for registering this with `{ signal }`

**Review by Rajdeepc — COMMENTED** [review]
Well structured PR. Pointing out some edge case and corner case testing. Then we are good to go!

1. Test for Shift+Tab: Ensure reverse tabbing `(Shift+Tab)` out of overlay also triggers delayed close.
2. Edge Cases: If overlays can contain nested overlays, test focus/close logic for multi-level overlays.

File: packages/overlay/src/HoverController.ts [hunk 7]
```diff
@@ -33,9 +33,21 @@ export class HoverController extends InteractionController {
 
     pointerentered = false;
 
+    overlayFocusedin = false;
```
Comment (Rajdeepc) [inline packages/overlay/src/HoverController.ts]: nit: You can consider renaming it to `overlayFocused` for clarity. These are minor and context dependent but clear names helps future maintainers.

Comment (cdransf) [reply packages/overlay/src/HoverController.ts]: Great call! Went ahead and updated the state booleans for clarity. ✨

**Comment by cdransf** [comment]
> Well structured PR. Pointing out some edge case and corner case testing. Then we are good to go!
> 
>     1. Test for Shift+Tab: Ensure reverse tabbing `(Shift+Tab)` out of overlay also triggers delayed close.
> 
>     2. Edge Cases: If overlays can contain nested overlays, test focus/close logic for multi-level overlays.

Went ahead and added these as well!

**Comment by Rajdeepc** [comment]
@cdransf There are some overlay tests failing due to this change. Can you please check this. Let me know if you would like to discuss any patterns.

**Comment by cdransf** [comment]
> @cdransf There are some overlay tests failing due to this change. Can you please check this. Let me know if you would like to discuss any patterns.

Got it! Everything is passing now, but I'd be happy to update things if they can be improved. I updated the golden hash as well since it didn't have a baseline to compare against for the VRTs.

File: packages/overlay/src/HoverController.ts [hunk 8]
```diff
@@ -27,15 +27,27 @@ export class HoverController extends InteractionController {
 
     private elementIds: string[] = [];
 
-    focusedin = false;
+    targetFocused = false;
```
Comment (caseyisonit) [inline packages/overlay/src/HoverController.ts]: can we make these private since they are only used and accessed within this controller?
```suggestion
    private targetFocused = false;
```

File: packages/overlay/src/HoverController.ts [hunk 9]
```diff
@@ -27,15 +27,27 @@ export class HoverController extends InteractionController {
...
 
     private hoverTimeout?: ReturnType<typeof setTimeout>;
 
-    pointerentered = false;
+    hovering = false;
+
+    overlayFocused = false;
```
Comment (caseyisonit) [inline packages/overlay/src/HoverController.ts]:
```suggestion
    private hovering = false;

    private overlayFocused = false;
```

File: packages/overlay/src/HoverController.ts [hunk 10]
```diff
@@ -75,11 +73,15 @@ export class HoverController extends InteractionController {
         this.doFocusleave();
     }
 
-    handleTargetPointerenter(): void {
+    protected clearCloseTimeout(): void {
```
Comment (rubencarvalho) [inline packages/overlay/src/HoverController.ts]: nit: curious why `protected` instead of `private`. But given this isn’t an exported public controller, I think it’s fine as-is; though if we wanted consistency, we’d probably update all methods and properties.

File: packages/overlay/src/HoverController.ts [hunk 11]
```diff
@@ -184,9 +229,9 @@ export class HoverController extends InteractionController {
     }
 
     override initOverlay(): void {
-        if (!this.abortController) {
-            return;
-        }
+        // Clean up listeners if they've already been bound
+        this.abortController?.abort();
+        this.abortController = new AbortController();
         const { signal } = this.abortController;
```
Comment (rubencarvalho) [inline packages/overlay/src/HoverController.ts]: I think we should not be aborting here, all the target element's event listeners are removed as soon as the overlay is initialized.

Comment (rubencarvalho) [reply packages/overlay/src/HoverController.ts]: Maybe you can create 2 AbortControllers, so they don't interefere with each other.

Comment (cdransf) [reply packages/overlay/src/HoverController.ts]: Just reverted this change. 👍🏻

File: packages/overlay/src/HoverController.ts (hunk 10 above)
Comment (cdransf) [reply packages/overlay/src/HoverController.ts]: Updated these additions to private. ✨ 

Should we make _all_ methods private?

**Review by caseyisonit — APPROVED** [review]
🚢 IT
